### ISSUE TITLE
bug(typescript): incorrect diagnostic "invalid function parameter"

### ISSUE THREAD

**Issue by vegerot** [quick-lint/quick-lint-js#929]
minimal(ish) repro:

```ts
export const A = (
  b: boolean,
  c: Record<string, any>
) => (error: GraphQLError): GraphQLFormattedError<Record<string, any>> => {
  const { code } = console.log(error) as any;
  if (code && !non2XXGraphQLErrors.has(code)) {
    c.graphqlError('/graphql');
  }
  if (!b || !code) {
    return error;
  }


  return error;
};
```

expected: 👌 (besides for builtin TypeScript utility types which haven't been implemented)

Actual:

<img width="2053" alt="image" src="https://user-images.githubusercontent.com/5297725/215625813-a945b3aa-5efd-414e-bc42-9cb0dcf16f2a.png">

```sh
❯ quick-lint-js --language=experimental-typescript src/server.ts
src/server.ts:577:12: error: unexpected ':' in expression; did you mean 'as'? [E0254]
src/server.ts:574:18: error: missing parentheses around parameter [E0255]
src/server.ts:577:27: note: TypeScript type annotation requires parentheses [E0255]
src/server.ts:574:18: error: invalid function parameter [E0151]
src/server.ts:151:24: warning: use of undeclared type: Record [E0214]
src/server.ts:491:20: warning: use of undeclared type: Record [E0214]
src/server.ts:499:29: warning: use of undeclared type: Partial [E0214]
src/server.ts:577:7: warning: use of undeclared variable: error [E0057]
src/server.ts:577:51: warning: use of undeclared type: Record [E0214]
src/server.ts:578:38: warning: use of undeclared variable: error [E0057]
src/server.ts:580:5: warning: use of undeclared variable: c [E0057]
src/server.ts:582:8: warning: use of undeclared variable: b [E0057]
src/server.ts:583:12: warning: use of undeclared variable: error [E0057]
src/server.ts:587:10: warning: use of undeclared variable: error [E0057]
src/server.ts:597:15: warning: use of undeclared type: Record [E0214]
src/server.ts:605:36: warning: use of undeclared type: Partial [E0214]
src/server.ts:738:18: warning: use of undeclared variable: formatError [E0057]
src/server.ts:861:45: warning: use of undeclared type: Record [E0214]
```

quick-lint version: quick-lint-js version 2.10.0 (d81893eaf4)

**Comment by strager** [comment]
There's a syntax ambiguity when there's a `:` in an expression arrow function's body, and quick-lint-js handles it wrong in your case:
```
//           arrow function
//           vvvvvvvvvvvvvvvvvvvvvvvv
       () => (somevar) : thing => { };
true ? () => (somevar) : thing => {};
//     ^^^^^^^^^^^^^^^   ^^^^^^^^^^^^
//     true branch       false branch
```
quick-lint-js currently says that [`:` is not part of the arrow function's expression](https://github.com/quick-lint/quick-lint-js/blob/d81893eaf434b447799ac5526f9ccd12f9439c4d/src/quick-lint-js/fe/parse-expression.cpp#L2472). This works for the `? :` case. However, in your case, quick-lint-js sees `() => (somevar)` as its own expression followed by an unexpected `:`. (Then quick-lint-js parses a type after `:`, leading to more confusing diagnostics.)

I need to change quick-lint-js' arrow function body parsing to only disregard `:` if we're in a `? :` expression.

**Comment by vegerot** [comment]
Makes sense.  Thank you!

**Comment by strager** [comment]
Fixed in commit e4b6abb7fcfd23d600fecf9556ae34ae7167176c.
